Core/include/Acts/TrackFitting/GlobalChiSquareFitter.hpp

```diff
@@ -391,38 +397,108 @@ class Gx2Fitter {
+          size_t currentTrackIndex = Acts::MultiTrajectoryTraits::kInvalid;
+          if (nUpdate == 0) {
+            ACTS_VERBOSE("   processSurface: nUpdate == 0 decision");
+
+            // Add a <mask> TrackState entry multi trajectory. This allocates
+            // storage for all components, which we will set later.
+            currentTrackIndex =
+                fittedStates.addTrackState(mask, result.lastTrackIndex);
+          } else {
+            ACTS_VERBOSE("   processSurface: nUpdate > 0 decision");
+
+            if (result.lastTrackIndex ==
+                Acts::MultiTrajectoryTraits::kInvalid) {
+              currentTrackIndex = 0;
+              ACTS_VERBOSE("   processSurface: currentTrackIndex (kInv->0) = "
+                           << currentTrackIndex);
+            } else if (result.lastTrackIndex < fittedStates.size() - 1) {
+              currentTrackIndex = result.lastTrackIndex + 1;
+              ACTS_VERBOSE("   processSurface: currentTrackIndex (n+1) = "
+                           << currentTrackIndex);
+            } else {
+              // Add a <mask> TrackState entry multi trajectory. This allocates
+              // storage for all components, which we will set later.
+              currentTrackIndex =
+                  fittedStates.addTrackState(mask, result.lastTrackIndex);
+              ACTS_VERBOSE("   processSurface: currentTrackIndex (ADD NEW)= "
+                           << currentTrackIndex);
+            }
+          }
 
           // now get track state proxy back
-          auto trackStateProxy = fittedStates.getTrackState(newTrackIndex);
-          trackStateProxy.setReferenceSurface(surface->getSharedPtr());
-          // assign the source link to the track state
-          trackStateProxy.setUncalibratedSourceLink(sourcelink_it->second);
-
-          // Fill the track state
-          trackStateProxy.predicted() = std::move(boundParams.parameters());
+          typename traj_t::TrackStateProxy trackStateProxy =
+              fittedStates.getTrackState(currentTrackIndex);
+
+          // Set the trackStateProxy components with the state from the ongoing
+          // propagation
+          {
+            trackStateProxy.setReferenceSurface(surface->getSharedPtr());
+            // Bind the transported state to the current surface
+            auto res = stepper.boundState(state.stepping, *surface, false,
+                                          freeToBoundCorrection);
+            if (!res.ok()) {
+              result.result = res.error();
+              return;
+            }
+            auto& [boundParams, jacobian, pathLength] = *res;
+
+            // Fill the track state
+            trackStateProxy.predicted() = std::move(boundParams.parameters());
+            if (boundParams.covariance().has_value()) {
+              trackStateProxy.predictedCovariance() =
+                  std::move(*boundParams.covariance());
```

Comment:
I don't think in these cases here, the `move` does anything. At the end of the day, the compiler has to do a copy here no matter what. I think for clarity, it might be worth removing the extra `move`s.